cmd/kritis/signer/server.go

```diff
@@ -0,0 +1,228 @@
+	}
+	return response, nil
+}
+
+// implements `/event` handles a container analysis occurrence notification
+func containerAnalysisEvent(w http.ResponseWriter, r *http.Request) {
+	var m PubSubMessage
+	var event ContainerAnalysisOccurrenceEvent
+	err := json.NewDecoder(r.Body).Decode(&m)
+	if err != nil {
+		http.Error(w, "could not unmarshal the pubsub message", http.StatusBadRequest)
+		return
+	}
+
+	err = json.Unmarshal(m.Message.Data, &event)
+	if err != nil {
+		http.Error(w, "could not unmarshal container analysis occurrence event from pubsub message", http.StatusBadRequest)
+		return
+	}
+
+	if event.Kind != "DISCOVERY" {
+		http.Error(w, "ignoring non-DISCOVERY occurrence", http.StatusOK)
+		return
+	}
+
+	occurrence, err := GetOccurrence(event)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	status := occurrence.GetDiscovered().GetDiscovered().GetAnalysisStatus()
+	if status != discovery.Discovered_FINISHED_SUCCESS {
+		http.Error(w, fmt.Sprintf("ignoring DISCOVERY occurrence in status %s", string(status)), http.StatusOK)
+		return
+	}
+	image := occurrence.Resource.Uri
+	if strings.HasPrefix(image, "https://") {
+		image = image[8:]
+	}
+	request := SignRequest{Image: image}
+	doCheckAndSign(w, request, http.StatusOK)
+}
+
+// implements `/check-only`
+func check(w http.ResponseWriter, r *http.Request) {
+	var request SignRequest
+	var response SignResponse
+
+	err := json.NewDecoder(r.Body).Decode(&request)
+	if err != nil {
+		response.Status = Failed
+		response.Message = fmt.Sprintf("failed to decode body, %s", err)
+		WriteResponse(w, response, http.StatusBadRequest)
+		return
+	}
+	response.Image = request.Image
+
+	violations, err := DoCheck(request.Image)
+	if err != nil {
+		response.Status = Failed
+		response.Message = err.Error()
+		WriteResponse(w, response, http.StatusBadRequest)
+		return
+	}
+	if violations != nil {
+		response.Status = Failed
+		response.Violations = violations
+		WriteResponse(w, response, http.StatusUnprocessableEntity)
+	} else {
+		response.Status = Ok
+		WriteResponse(w, response, http.StatusOK)
+	}
+}
+
+// check and sign is called either directly via /check-and-sign or indirect via
+// the /event notification.
+func doCheckAndSign(w http.ResponseWriter, request SignRequest, violationStatus int) {
+	response := SignResponse{Image: request.Image}
+
+	violations, err := DoCheck(request.Image)
+	if err != nil {
+		response.Status = Failed
+		response.Message = err.Error()
+		WriteResponse(w, response, http.StatusBadRequest)
+		return
+	}
+
+	if violations != nil {
+		response.Status = NotSigned
+		response.Message = fmt.Sprintf("image violates policy")
+		response.Violations = violations
+		WriteResponse(w, response, violationStatus)
+		return
+	}
+
+	err = DoSign(request.Image)
+	if err != nil {
+		response.Status = Failed
+		response.Message = err.Error()
+		WriteResponse(w, response, http.StatusInternalServerError)
```

Comment:
Should this really be an internal error?  Also, it seems it should be consistent with the error in L172.

Reply:
Line 172 is indeed incorrect. I fixed that. It is an internal server error as we failed to do what we were requested to do and the caller has not way of fixing that. We need to do that.